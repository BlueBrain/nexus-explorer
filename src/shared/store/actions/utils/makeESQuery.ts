import { ResourceBoardList } from '../../../containers/ResourceListBoardContainer';

const managementSchemas = [
  'https://bluebrain.github.io/nexus/schemas/resolver.json',
  'https://bluebrain.github.io/nexus/schemas/storage.json',
  'https://bluebrain.github.io/nexus/schemas/view.json',
];

interface TermFilter {
  term: { [termKey: string]: string };
}

// TODO break out into library
export const makeESQuery = (query?: ResourceBoardList['query']) => {
  const sort = [
    {
      _createdAt: 'desc',
    },
    {
      '@id': 'desc',
    },
  ];

  if (query) {
    const must = [];
    const mustNot: any[] = []; // too lazy to type all of the possible query types!
    if (Object.keys(query.filters).length) {
      Object.keys(query.filters).forEach(key => {
        // exception to make filtering easier
        // because management resources are many
        if (key === 'showManagementResources') {
          const should: TermFilter[] = []; // OR === Should
          managementSchemas.forEach(schema => {
            should.push({
              term: { _constrainedBy: schema },
            });
          });
          if (!!query.filters[key]) {
            // showManagementResoruces is true, add them as
            // a must match to the filter
            must.push({
              bool: {
                should,
              },
            });
          } else {
            // otherwise add as mustNot match to filter them out
            mustNot.push({
              bool: {
                should,
              },
            });
          }
          // Otherwise we ignore this value, as it's
          // not a real filterable value
          // just a convenience label
          return;
        }
        // any other keys inside the filter object
        // are legitimate filterable keys
        must.push({
          term: { [key]: query.filters[key] },
        });
      });
    }
    if (query.textQuery) {
      must.push({
        // from example: https://github.com/BlueBrain/nexus-kg/blob/master/src/test/resources/search/query-schema-rev-createdBy-q.json#L10-L12
        match: {
          _all_fields: query.textQuery,
        },
      });
    }

    if (must.length >= 1 || mustNot.length >= 1) {
      return {
        sort,
        query: {
          bool: {
            must,
            must_not: mustNot,
          },
        },
      };
    }
    if (must.length === 0) {
      return {
        sort,
      };
    }
    return {
      sort,
      query: {
        ...must[0],
      },
    };
  }
  return {};
};
