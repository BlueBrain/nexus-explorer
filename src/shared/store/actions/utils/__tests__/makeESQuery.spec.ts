import { makeESQuery } from '../makeESQuery';

describe('makeESQuery', () => {
  it('should return an empty object if no query prop is provided', () => {
    expect(makeESQuery()).toEqual({});
  });

  describe('text query', () => {
    it('should return a must clause with a query string', () => {
      expect(
        makeESQuery({
          filters: {
            _deprecated: false,
            showManagementResources: false,
          },
          textQuery: 'banana',
        })
      ).toMatchSnapshot();
    });
  });

  describe('deprecated', () => {
    it('should return a must clause with just the term query', () => {
      expect(
        makeESQuery({
          filters: {
            _deprecated: false,
            showManagementResources: false,
          },
        })
      ).toMatchSnapshot();
    });
  });

  describe('_constrainedBy', () => {
    it('should return a must clause with a term query for both _deprecated and _constrainedBy', () => {
      expect(
        makeESQuery({
          filters: {
            _constrainedBy: 'mySchema',
            _deprecated: false,
            showManagementResources: false,
          },
        })
      ).toMatchSnapshot();
    });
  });

  describe('@type', () => {
    it('should return a must clause with a term query for both _deprecated and @types', () => {
      expect(
        makeESQuery({
          filters: {
            '@type': 'myType',
            _deprecated: true,
            showManagementResources: false,
          },
        })
      ).toMatchSnapshot();
    });
  });

  describe('everything at once', () => {
    it('should return a must clause with the kitchen sink', () => {
      expect(
        makeESQuery({
          filters: {
            '@type': 'myType',
            _constrainedBy: 'mySchemas',
            _deprecated: true,
            showManagementResources: false,
          },
          textQuery: 'Banana',
        })
      ).toMatchSnapshot();
    });
  });
});
