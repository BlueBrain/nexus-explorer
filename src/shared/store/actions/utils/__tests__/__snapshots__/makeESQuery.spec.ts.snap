// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeESQuery @type should return a must clause with a term query for both _deprecated and @types 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "@type": "myType",
          },
        },
        Object {
          "term": Object {
            "_deprecated": true,
          },
        },
      ],
      "must_not": Array [
        Object {
          "bool": Object {
            "should": Array [
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/resolver.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/storage.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/view.json",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`makeESQuery _constrainedBy should return a must clause with a term query for both _deprecated and _constrainedBy 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "_constrainedBy": "mySchema",
          },
        },
        Object {
          "term": Object {
            "_deprecated": false,
          },
        },
      ],
      "must_not": Array [
        Object {
          "bool": Object {
            "should": Array [
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/resolver.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/storage.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/view.json",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`makeESQuery deprecated should return a must clause with just the term query 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "_deprecated": false,
          },
        },
      ],
      "must_not": Array [
        Object {
          "bool": Object {
            "should": Array [
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/resolver.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/storage.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/view.json",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`makeESQuery everything at once should return a must clause with the kitchen sink 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "@type": "myType",
          },
        },
        Object {
          "term": Object {
            "_constrainedBy": "mySchemas",
          },
        },
        Object {
          "term": Object {
            "_deprecated": true,
          },
        },
        Object {
          "query_string": Object {
            "query": "Banana~",
          },
        },
      ],
      "must_not": Array [
        Object {
          "bool": Object {
            "should": Array [
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/resolver.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/storage.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/view.json",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`makeESQuery text query should return a must clause with a query string 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "_deprecated": false,
          },
        },
        Object {
          "query_string": Object {
            "query": "banana~",
          },
        },
      ],
      "must_not": Array [
        Object {
          "bool": Object {
            "should": Array [
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/resolver.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/storage.json",
                },
              },
              Object {
                "term": Object {
                  "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/view.json",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;
